plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/transport/simulate/ParsedSimulateRequest.java

```diff
@@ -0,0 +1,110 @@
+        return documents;
+    }
+
+    public boolean isVerbose() {
+        return verbose;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        ParsedSimulateRequest that = (ParsedSimulateRequest) o;
+        return Objects.equals(verbose, that.verbose) &&
+                Objects.equals(documents, that.documents) &&
+                Objects.equals(pipeline, that.pipeline);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(documents, pipeline, verbose);
+    }
+
+    public static class Parser {
+        private static final Pipeline.Factory PIPELINE_FACTORY = new Pipeline.Factory();
+        public static final String SIMULATED_PIPELINE_ID = "_simulate_pipeline";
+
+        private List<Data> parseDocs(List<Map<String, Object>> docs) {
+            List<Data> dataList = new ArrayList<>();
+            for (Map<String, Object> dataMap : docs) {
+                Map<String, Object> document = ConfigurationUtils.readMap(dataMap, Fields.SOURCE);
+                Data data = new Data(ConfigurationUtils.readStringProperty(dataMap, Fields.INDEX),
+                        ConfigurationUtils.readStringProperty(dataMap, Fields.TYPE),
+                        ConfigurationUtils.readStringProperty(dataMap, Fields.ID),
+                        document);
+                dataList.add(data);
+            }
+            return dataList;
+        }
+
+        public ParsedSimulateRequest parseWithPipelineId(String pipelineId, Map<String, Object> config, boolean verbose, PipelineStore pipelineStore) {
+            if (pipelineId == null) {
+                throw new IllegalArgumentException("param [pipeline] is null");
+            }
+            Pipeline pipeline = pipelineStore.get(pipelineId);
+            List<Data> dataList = parseDocs(ConfigurationUtils.readList(config, Fields.DOCS));
+            return new ParsedSimulateRequest(pipeline, dataList, verbose);
+
+        }
+
+        public ParsedSimulateRequest parse(Map<String, Object> config, boolean verbose, PipelineStore pipelineStore) throws IOException {
+            Map<String, Object> pipelineConfig = ConfigurationUtils.readOptionalMap(config, Fields.PIPELINE);
+            Pipeline pipeline = PIPELINE_FACTORY.create(SIMULATED_PIPELINE_ID, pipelineConfig, pipelineStore.getProcessorFactoryRegistry());
+            List<Data> dataList = parseDocs(ConfigurationUtils.readList(config, Fields.DOCS));
+            return new ParsedSimulateRequest(pipeline, dataList, verbose);
+        }
+
+        static final class Fields {
+            static final String PIPELINE = "pipeline";
+            static final String DOCS = "docs";
+            static final String SOURCE = "_source";
+            static final String INDEX = "_index";
+            static final String TYPE = "_type";
+            static final String ID = "_id";
```

Comment:
seems like some of these contants could be moved to some common place, made public and re-used in other places